crates/catalog/inmemory/src/catalog.rs

```diff
@@ -0,0 +1,1511 @@
+    async fn get_namespace(&self, namespace_ident: &NamespaceIdent) -> Result<Namespace> {
+        let root_namespace_state = self.root_namespace_state.lock().await;
+
+        let namespace = Namespace::with_properties(
+            namespace_ident.clone(),
+            root_namespace_state
+                .get_properties(namespace_ident)?
+                .clone(),
+        );
+
+        Ok(namespace)
+    }
+
+    /// Check if namespace exists in catalog.
+    async fn namespace_exists(&self, namespace_ident: &NamespaceIdent) -> Result<bool> {
+        let guarded_namespaces = self.root_namespace_state.lock().await;
+
+        Ok(guarded_namespaces.namespace_exists(namespace_ident))
+    }
+
+    /// Update a namespace inside the catalog.
+    ///
+    /// # Behavior
+    ///
+    /// The properties must be the full set of namespace.
+    async fn update_namespace(
+        &self,
+        namespace_ident: &NamespaceIdent,
+        properties: HashMap<String, String>,
+    ) -> Result<()> {
+        let mut root_namespace_state = self.root_namespace_state.lock().await;
+
+        root_namespace_state.replace_properties(namespace_ident, properties)
+    }
+
+    /// Drop a namespace from the catalog.
+    async fn drop_namespace(&self, namespace_ident: &NamespaceIdent) -> Result<()> {
+        let mut root_namespace_state = self.root_namespace_state.lock().await;
+
+        root_namespace_state.remove_existing_namespace(namespace_ident)
+    }
+
+    /// List tables from namespace.
+    async fn list_tables(&self, namespace_ident: &NamespaceIdent) -> Result<Vec<TableIdent>> {
+        let root_namespace_state = self.root_namespace_state.lock().await;
+
+        let table_names = root_namespace_state.list_tables(namespace_ident)?;
+        let table_idents = table_names
+            .into_iter()
+            .map(|table_name| TableIdent::new(namespace_ident.clone(), table_name.clone()))
+            .collect_vec();
+
+        Ok(table_idents)
+    }
+
+    /// Create a new table inside the namespace.
+    async fn create_table(
+        &self,
+        namespace_ident: &NamespaceIdent,
+        table_creation: TableCreation,
+    ) -> Result<Table> {
+        let mut root_namespace_state = self.root_namespace_state.lock().await;
+
+        let table_name = table_creation.name.clone();
+        let table_ident = TableIdent::new(namespace_ident.clone(), table_name);
+        let table_exists = root_namespace_state.table_exists(&table_ident)?;
+
+        if table_exists {
+            Err(Error::new(
+                ErrorKind::Unexpected,
+                format!(
+                    "Cannot create table {:?}. Table already exists",
+                    &table_ident
+                ),
+            ))
+        } else {
+            let (table_creation, location) = match table_creation.location.clone() {
+                Some(location) => (table_creation, location),
+                None => {
+                    let tmp_dir = TempDir::new()?;
```

Comment:
I personally believe we should not depend on `TempDir`. As mentioned in the previous comment, the `file_io` could be memory or s3. 

If `location` is empty, we can generate one for user.

Reply:
Agreed; if we're going to accept arbitrary `FileIO` instances, it doesn't make sense to use `TempDir` anymore here. I've refactored the code so that you can configure a default table location when you initialize the `InMemoryCatalog` that will be used if no location is provided in `TableCreation` instances.